### PR TITLE
[vulkan] Refactor vkEnumerate methods

### PR DESCRIPTION
**Pull request by poweifeng**
Use templates to help clean up VK property/layer enumerations.

### CODE REVIEW

File: filament/backend/src/vulkan/VulkanUtility.h [hunk 1]
```diff
@@ -66,6 +68,46 @@ bool equivalent(const VkExtent2D& a, const VkExtent2D& b);
...
+}
+
+template <typename InTypeA, typename InTypeB, typename OutType>
+utils::FixedCapacityVector<OutType> enumerate(
+        VkResult (*func)(InTypeA, InTypeB, uint32_t*, OutType*), InTypeA inDataA, InTypeB inDataB) {
+    EXPAND_ENUM_ARGS(inDataA, inDataB);
+}
+
+#undef EXPAND_ENUM
+#undef EXPAND_ENUM_NO_ARGS
```
Comment (bejado) [inline filament/backend/src/vulkan/VulkanUtility.h:108]:
```suggestion
#undef EXPAND_ENUM_NO_ARGS
```

File: filament/backend/src/vulkan/VulkanUtility.h [hunk 2]
```diff
@@ -66,6 +68,46 @@ bool equivalent(const VkExtent2D& a, const VkExtent2D& b);
...
+// non-variadic parts, this breaks the template type matching logic. Hence, we use a macro approach
+// here.
+#define EXPAND_ENUM(...)\
+    uint32_t size = 0;\
+    VkResult result = func(__VA_ARGS__, nullptr);\
+    ASSERT_POSTCONDITION(result == VK_SUCCESS && size > 0, "enumerate size error");\
+    utils::FixedCapacityVector<OutType> ret(size);\
+    result = func(__VA_ARGS__, ret.data());\
+    ASSERT_POSTCONDITION(result == VK_SUCCESS, "enumerate error");\
+    return move(ret);
```
Comment (bejado) [inline filament/backend/src/vulkan/VulkanUtility.h:86]: Prefix with `std::`?

**Review by bejado — APPROVED** [review]

File: filament/backend/src/vulkan/VulkanUtility.h (hunk 2 above)
Comment (poweifeng) [reply filament/backend/src/vulkan/VulkanUtility.h]: done

File: filament/backend/src/vulkan/VulkanUtility.h (hunk 1 above)
Comment (poweifeng) [reply filament/backend/src/vulkan/VulkanUtility.h]: done

**Pull request merged**